ALICE3/TableProducer/onTheFlyTOFPID.cxx

```diff
@@ -0,0 +1,279 @@
+  TRandom3 lPRNG;
+
+  // for handling basic QA histograms if requested
+  HistogramRegistry histos{"Histos", {}, OutputObjHandlingPolicy::AnalysisObject};
+
+  void init(o2::framework::InitContext& initContext)
+  {
+    lPRNG.SetSeed(0); // fully randomize
+
+    if (doQAplots) {
+      const AxisSpec axisMomentum{(int)80, 0.0f, +4.0f, "#it{p} (GeV/#it{c})"};
+      const AxisSpec axisVelocity{(int)110, 0.0f, +1.1f, "Measured #beta"};
+      histos.add("h2dVelocityVsMomentumInner", "h2dVelocityVsMomentumInner", kTH2F, {axisMomentum, axisVelocity});
+      histos.add("h2dVelocityVsMomentumOuter", "h2dVelocityVsMomentumOuter", kTH2F, {axisMomentum, axisVelocity});
+    }
+  }
+
+  template <typename mcParticleType>
+  void convertMCParticleToO2Track(mcParticleType& particle, o2::track::TrackParCov& o2track)
+  {
+    // FIXME: this is a fundamentally important piece of code.
+    // It could be placed in a utility file instead of here.
+    std::array<float, 3> xyz = {static_cast<float>(particle.vx()), static_cast<float>(particle.vy()), static_cast<float>(particle.vz())};
+    std::array<float, 3> ptetaphi = {static_cast<float>(particle.pt()), static_cast<float>(particle.eta()), static_cast<float>(particle.phi())};
+    auto pdgInfo = pdg->GetParticle(particle.pdgCode());
+    int charge = 0;
+    if (pdgInfo != nullptr)
+      charge = pdgInfo->Charge();
+    std::array<float, 5> params;
+    std::array<float, 15> covm = {
+      0.,
+      0., 0.,
+      0., 0., 0.,
+      0., 0., 0., 0.,
+      0., 0., 0., 0., 0.};
+    float s, c, x;
+    o2::math_utils::sincos(ptetaphi[2], s, c);
+    o2::math_utils::rotateZInv(xyz[0], xyz[1], x, params[0], s, c);
+    params[1] = xyz[2];
+    params[2] = 0.; // since alpha = phi
+    auto theta = 2. * atan(exp(-ptetaphi[1]));
+    params[3] = 1. / tan(theta);
+    params[4] = charge / ptetaphi[0];
+
+    // Initialize TrackParCov in-place
+    new (&o2track)(o2::track::TrackParCov)(x, ptetaphi[2], params, covm);
+  }
+
+  Double_t TrackLength(o2::track::TrackParCov track, Double_t lX0, Double_t lX1, Double_t lMagneticField)
+  {
+    std::array<float, 3> lPointN;
+    std::array<float, 3> lPointNplus;
+    Double_t lLength = 0.0;
+    track.propagateTo(lX0, lMagneticField);
+    for (Int_t iStep = 1; iStep < lNStepsLIntegrator; iStep++) {
+      track.getXYZGlo(lPointN);
+      Float_t lPosition = lX0 + (lX1 - lX0) * ((Float_t)(iStep)) / ((Float_t)(lNStepsLIntegrator - 1));
+      track.propagateTo(lPosition, lMagneticField);
+      track.getXYZGlo(lPointNplus);
+      lLength += std::hypot(lPointNplus[0] - lPointN[0], lPointNplus[1] - lPointN[1], lPointNplus[2] - lPointN[2]);
+    }
+    return lLength;
+  }
+
+  Double_t Velocity(Double_t lMomentum, Double_t lMass)
+  {
+    // Momentum p and mass m -> returns speed in centimeters per picosecond
+    // Useful for TOF calculations
+    Double_t lA = TMath::Power(lMomentum / lMass, 2);
+    return 0.0299792458 * TMath::Sqrt(lA / (1 + lA));
```

Comment:
The 0.0299792458 constant needs some documentation.

Reply:
I would suggest using `std::` functions.

Reply:
done!